### ISSUE TITLE
Windows minion dies on startup if cannot reach master

### ISSUE THREAD

**Issue by Ch3LL** [saltstack/salt#33277]
### Description of Issue/Question
On startup of a windows minion if the master is not available the windows minion will die and not attempt to re-authenticate with the master.

### Setup
A windows minion who's key as previously been accepted by the master. This occurs in both single and multi-master mode.

/etc/salt/minion

`master: 192.168.50.10`

or 

```
master:
  - 192.168.50.10
  - 192.168.50.11
```

### Steps to Reproduce Issue
1. Setup a windows minion and point to master
2. Accept the key on the master
3. Stop the master and the minion
4. Start the minion and see the minion attempt to communicate with the down master.
5. the minion will then die after its attempted to communicate with the down master. 

Here is the debug log output:

```
PS C:\salt> .\salt-minion.bat -ldebug
[DEBUG   ] Reading configuration from c:\salt\conf\minion
[DEBUG   ] Including configuration from 'c:\salt\conf\minion.d\_schedule.conf'
[DEBUG   ] Reading configuration from c:\salt\conf\minion.d\_schedule.conf
[DEBUG   ] Configuration file path: c:\salt\conf\minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[INFO    ] Setting up the Salt Minion "win2012"
[DEBUG   ] Created pidfile: c:\salt\var\run\salt-minion.pid
[DEBUG   ] Reading configuration from c:\salt\conf\minion
[DEBUG   ] Including configuration from 'c:\salt\conf\minion.d\_schedule.conf'
[DEBUG   ] Reading configuration from c:\salt\conf\minion.d\_schedule.conf
[INFO    ] Creating minion process manager
[WARNING ] IMPORTANT: Do not use md5 hashing algorithm! Please set "hash_type" to sha256 in Salt Minion config!
[INFO    ] The Salt Minion is starting up
[INFO    ] Minion is starting as user 'ch3ll'
[DEBUG   ] AsyncEventPublisher PUB socket URI: 4510
[DEBUG   ] AsyncEventPublisher PULL socket URI: 4511
[INFO    ] Starting pull socket on 4511
[DEBUG   ] Minion 'win2012' trying to tune in
[DEBUG   ] sync_connect_master
[CRITICAL] 'master_type' set to 'failover' but 'retry_dns' is not 0. Setting 'retry_dns' to 0 to failover to the next m
ster on DNS errors.
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Initializing new AsyncAuth for ('c:\\salt\\conf\\pki\\minion', 'win2012', 'tcp://192.168.50.10:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (2296)
[DEBUG   ] Setting zmq_reconnect_ivl to '2296ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Process Manager starting!
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('c:\\salt\\conf\\pki\\minion', 'win2012', 'tcp://192.168.50.10:4
06', 'clear')
[DEBUG   ] SaltReqTimeoutError, retrying. (1/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (2/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (3/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (4/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (5/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (6/7)
[DEBUG   ] SaltReqTimeoutError, retrying. (7/7)
[INFO    ] Master 192.168.50.10 could not be reached, trying next master (if any)
[ERROR   ] No master could be reached or all masters denied the minions connection attempt.
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
SaltClientError: Attempt to authenticate with the salt master failed with timeout error
Traceback (most recent call last):
  File "C:\salt\bin\Scripts\salt-minion", line 26, in <module>
    salt_minion()
  File "C:\salt\bin\lib\site-packages\salt\scripts.py", line 112, in salt_minion
    minion.start()
  File "C:\salt\bin\lib\site-packages\salt\cli\daemons.py", line 349, in start
    self.minion.tune_in()
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 1961, in tune_in
    self.sync_connect_master()
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 896, in sync_connect_master
    raise six.reraise(*future_exception)
  File "C:\salt\bin\lib\site-packages\tornado\gen.py", line 1014, in run
    yielded = self.gen.throw(*exc_info)
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 905, in connect_master
    master, self.pub_channel = yield self.eval_master(self.opts, self.timeout, self.safe)
  File "C:\salt\bin\lib\site-packages\tornado\gen.py", line 1008, in run
    value = future.result()
  File "C:\salt\bin\lib\site-packages\tornado\concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "C:\salt\bin\lib\site-packages\tornado\gen.py", line 1014, in run
    yielded = self.gen.throw(*exc_info)
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 540, in eval_master
    raise last_exc  # pylint: disable=E0702
SaltClientError: Attempt to authenticate with the salt master failed with timeout error
Traceback (most recent call last):
  File "C:\salt\bin\Scripts\salt-minion", line 26, in <module>
    salt_minion()
  File "C:\salt\bin\lib\site-packages\salt\scripts.py", line 112, in salt_minion
    minion.start()
  File "C:\salt\bin\lib\site-packages\salt\cli\daemons.py", line 349, in start
    self.minion.tune_in()
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 1961, in tune_in
    self.sync_connect_master()
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 896, in sync_connect_master
    raise six.reraise(*future_exception)
  File "C:\salt\bin\lib\site-packages\tornado\gen.py", line 1014, in run
    yielded = self.gen.throw(*exc_info)
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 905, in connect_master
    master, self.pub_channel = yield self.eval_master(self.opts, self.timeout, self.safe)
  File "C:\salt\bin\lib\site-packages\tornado\gen.py", line 1008, in run
    value = future.result()
  File "C:\salt\bin\lib\site-packages\tornado\concurrent.py", line 232, in result
    raise_exc_info(self._exc_info)
  File "C:\salt\bin\lib\site-packages\tornado\gen.py", line 1014, in run
    yielded = self.gen.throw(*exc_info)
  File "C:\salt\bin\lib\site-packages\salt\minion.py", line 540, in eval_master
    raise last_exc  # pylint: disable=E0702
salt.exceptions.SaltClientError: Attempt to authenticate with the salt master failed with timeout error
```

### Versions Report

```
PS C:\salt> .\salt-minion.bat --versions-report
Salt Version:
           Salt: 2016.3.0rc3

Dependency Versions:
           cffi: 1.6.0
       cherrypy: 5.3.0
       dateutil: 2.5.3
          gitdb: 0.6.4
      gitpython: 2.0.2
          ioflo: 1.5.3
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: 1.4.4
       M2Crypto: Not Installed
           Mako: 1.0.4
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.11 (v2.7.11:6d1b6a68f775, Dec  5 2015, 20:40:30) [MSC v.1500 64 bit (AMD64)]
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: 0.2.4
        Tornado: 4.3
            ZMQ: 4.1.2

System Versions:
           dist:
        machine: AMD64
        release: 8.1
         system: Windows
        version: 8.1 6.3.9600  Multiprocessor Free
```

Also tested this on windows 2012

**Comment by cachedout** [comment]
This is limited to 2016.3, right? Did you double-check to make sure this isn't also on 2015.8?

**Comment by cachedout** [comment]
Also, has this been confirmed as being Windows-only? I would like to know if the same things happen in Linux if one toggles the `multiprocessing` config option on the minion.

**Comment by Ch3LL** [comment]
@cachedout I will double-check 2015.8 really quick. But as far as i know this is only 2016.3

Yes this is windows-only. But I will test with multiprocessing real quick on linux  as well..

**Comment by cachedout** [comment]
@Ch3LL Cool. Keep me posted. :]

**Comment by Ch3LL** [comment]
@cachedout  Okay so this also occurring in 2015.8 as well. I tested on 2015.8.8 2015.8.9 and head of 2015.8 and still am seeing the same behavior. The cannot connect ot the master and then errors and kills the process.

When testing with linux and switching multiprocessing to True on 2016.3.0rc3 did NOT result in the minion dying, it simply kept attempting to re-auth with the master.

**Comment by cachedout** [comment]
@Ch3LL Yup, OK. That lines up with what I'd expect after looking at this code. @twangboy is investigating.

**Comment by cachedout** [comment]
@twangboy has a fix for this on 2016.3 in #33285 but will need a separate one for 2015.8.

**Comment by ssgward** [comment]
Closing. This fix has been verified on both 2015.8 and 2016.3
